Prediction markets challenge

phipsae opened the pull request:
Hi all,

Here comes the prediction markets challenge. 
I tested it a couple of times, went through the text, but for sure there is still some room for improvement :) 

Looking forward to your feedback!

ZakGriffith commented:
I did a pass on this and have a handful of readMe updates ready to PR in once this get's merged.  Looks very near done to me and running the challenge went smooth.

phipsae commented:
@ZakGriffith Thank you!

rin-st commented on extension/packages/nextjs/components/Header.tsx.args.mjs:
```diff
@@ -1,3 +1,19 @@
+export const menuIconImports = `import { BanknotesIcon, EyeIcon, UserIcon } from "@heroicons/react/24/outline";`;
+export const menuObjects = `{
```
not sure what's the problem for now but when trying `yarn vercel` I'm getting this error on build
<img width="1488" alt="image" src="https://github.com/user-attachments/assets/4fd0dc72-1153-454c-9798-afa31c559591" />

error message:
`Type error: Type '({ label: string; href: string; } | { label: string; href: string; icon: Element; } | undefined)[]' is not assignable to type 'HeaderMenuLink[]'`

rin-st replied:
oh, I resulting code in the header is
<img width="460" alt="image" src="https://github.com/user-attachments/assets/9af7e96f-74e9-46ac-8b5d-aa53030f7202" />

Looks like you need to remove that `,` after Oracle object

rin-st commented on extension/packages/nextjs/components/race/Race.tsx:
```diff
@@ -0,0 +1,261 @@
+              disabled={raceStarted}
+              className="px-4 py-2 rounded-md bg-green-500 hover:bg-green-600 text-white font-bold transition-colors"
+            >
+              Start Race
```
Is it possible to start race without clicking on reset? (just asking, for me it would be better)

phipsae replied:
You mean getting rid of the reset and only have the start button?

ZakGriffith replied:
I agree with this, the start button would restart the race if it has already concluded.

Another helpful feature would be if the race reset itself on the front end after a contract re-deploy.

rin-st reviewed (commented):
Thanks for a great challenge! It was fun testing it!

Added some comments, mostly typos

phipsae commented on README.md:
```diff
@@ -1,6 +1,1268 @@
+
+2. **Smooths price movements** early on. Since the token price is calculated as:
+
+   $$tokenPrice = i\_initialTokenValue * marketProbability$$
```
Where? I don't see it.

rin-st replied:
github markdown shows this
<img width="790" alt="image" src="https://github.com/user-attachments/assets/1989570a-7ec2-439b-81fc-331a5ee23850" />

phipsae replied:
Weird in preview (cursor) it works correctly, but whenever its pushed it doesn't. Do you know how the syntax has to be (that latex work)?

phipsae commented on README.md:
````diff
@@ -1,6 +1,1268 @@
+
+```
+yarn test --grep "Checkpoint5"
+```
````
You mean I should add again the functions add/remove with the modifier?

rin-st replied:
No, I think just mentioning that the modifier should be added to those functions is enough

phipsae replied:
I could add the sentence "Also we want to add it to the functions `addLiquidity` and `removeLiquidity`." Not sure if we need that since we already have a hint to that in Guiding question 3 "Which other existing functions should also use the predictionNotReported modifier?". 

Let me know :)

rin-st replied:
Ok, let's not change it for now and see if it will be questioned often in the challenge tg channel

Pabl0cks commented on README.md:
```diff
@@ -1,6 +1,1265 @@
+
+> ❗️In the current state your front-end is already implemented but the buttons of the different function will likely break since there is no implementation code within your smart contracts yet. But soon there will! 🙂
+
+![ch-6-lp1](extension/packages/nextjs/public/lp1.png)
+
+In **Checkpoint 5**, you'll bring the Oracle to life by implementing the logic for outcome reporting.
+
+![ch-6-oracle1](extension/packages/nextjs/public/oracle1.png)
+
+Finally, in **Checkpoints 7 to 9**, we shift our focus to the User, where you'll implement the functionalities: **buying, selling, and redeeming** outcome tokens.
+
+![ch-6-user1](extension/packages/nextjs/public/user1.png)
```
Maybe we are anticipating too many stuff from future Checkpoints in this Checkpoint 1?

phipsae replied:
How do you mean? Would you prefer to not give an outlook what will happen in the challenge?

Pabl0cks replied:
I think we might be going too far by showing three fullscreen screenshots there

phipsae replied:
I made them expandable

Pabl0cks commented on README.md:
```diff
@@ -1,6 +1,1265 @@
+
+<details markdown='1'><summary>Question 4</summary>
+
+How do we validate that `_initialYesProbability` is within the valid range of 0% to 100%?
```
Missing the `>`

Pabl0cks reviewed (commented):
GJ @phipsae, this was a huge challenge to tackle. A few notes after the first round of test:

- Why aren’t we using the same color styles and typography as the other challenges? Would be nice to keep things visually consistent unless we want to differentiate some challenges from others in the future for a specific reason (like different "families" of challenges for example)
- We’re using `create-eth@0.1.8` here, but the rest use `0.1.0`. Not sure what's the correct approach should be here, or if we need `0.1.8` for some specific feature. Just checking 🙏
- Overall, the Readme feels a bit too verbose, with a lot of background information. Maybe we could make it more concise and to the point in some of the Checkpoints? The tone also feels more academic/formal compared to other challenges, which tend to be more conversational and encouraging.
- Some ideas for the Readme
   - What if we move some of the background/theory into SRE articles or guides and link to them from the Readme? Alternatively, we could make those sections collapsible, so builders can choose how deep they want to go.
   - In many challenges, we reuse the homepage intro paragraphs as part of opening section of the Readme. It gives users a high-level overview of the challenge before diving into the details.
   - We're missing some of the "emoji game" and Goals / Side quests(if applicable) subsections that are common in other challenges.
   
Sorry to open so many questions, I think it's worth discussing together about them 🙌

phipsae commented:
Thanks for all the feedback! 

- What do exactly mean? The font? Missing of the green background? 
- Since it was built with 0.1.8 version of SE2, I think it's fine to go with 0.1.8. That way, we ensure compatibility even if a newer SE2 version comes out that might introduce breaking changes.
- Could you share a few examples of the parts that felt too verbose or that you didn’t like? For me, all of these sections are important (otherwise I wouldn’t have included them :)) They are there to help explain how a prediction market works. If we take them out, I don't know how they should solve it and then whats the point of the challenge?
- Which homepage do you mean? (Cannot find what you mean)
- Not a big fan of too many emojis, to me it distracts, but if we want we can add a ton

It would’ve been helpful to get this kind of feedback earlier, especially since the README has been out for a while and I put quite a bit of time into refining it.

rin-st commented on README.md:
```diff
@@ -1,6 +1,1268 @@
+
+Now you could head over to the oracle tab in the UI and report the outcome (after you watched the race of course:)) 🏎️ 🏁
+
+> 💡 Make sure you're connected with the correct oracle address — check `00_deploy_your_contract.ts` to find out which one is being used.
```
As I understand you should be connected on the frontend with the same address to be able to report the outcome, right? How user should connect to that address on the fronend?

Image right after this sentence
<img width="1139" alt="image" src="https://github.com/user-attachments/assets/7ec4b207-5210-49b6-aff7-742b1fab2f25" />

rin-st replied:
It was not needed in previous challenges, so probably some users don't know it. Up to you, it can just lead to additional questions in tg channel

Pabl0cks commented:
> What do exactly mean? The font? Missing of the green background?

I mean to the overall look and feel. When you check the rest of the challenges, you can tell they’re all part of the same product (background color, font, button corners and sizes, border colors, hover effects). This challenge’s look and feel seems like it belongs to a different product. I would understand if there was a design reason behind it, as I mentioned in my previous comment, or having some small tweaks because this challenge requires some extra visual components.

Having dark screenshots in the README just makes the difference even more obvious.

![challenges-overview](https://github.com/user-attachments/assets/4342afe3-7b48-4b44-b0f5-ad694928b5c0)

> Since it was built with 0.1.8 version of SE2, I think it's fine to go with 0.1.8. That way, we ensure compatibility even if a newer SE2 version comes out that might introduce breaking changes.

Ok!

> Which homepage do you mean? (Cannot find what you mean)

I mean the homepage of the challenge. 

Just wanted to share the idea that in some READMEs, we use these two paragraphs (marked in red) as the first and last paragraphs of the initial section.

You can check other challenges, like [challenge 1 readme](https://github.com/scaffold-eth/se-2-challenges/blob/challenge-1-decentralized-staking/README.md) and [challenge 1 homepage](https://github.com/scaffold-eth/se-2-challenges/blob/challenge-1-decentralized-staking/extension/packages/nextjs/app/page.tsx.args.mjs), for reference

<details><summary>See homepage screenshot with the 2 paragraphs</summary>

![image](https://github.com/user-attachments/assets/fb571814-8389-4cfd-b477-02066fe9c2b4)

</details>

> Could you share a few examples of the parts that felt too verbose or that you didn’t like? For me, all of these sections are important (otherwise I wouldn’t have included them :)) They are there to help explain how a prediction market works. If we take them out, I don't know how they should solve it and then whats the point of the challenge?

- This wall of text right at the top of the README feels like it might overwhelm builders before they even begin to engage. I think we could format it in a more digestible way, visual elements - formatting to not be plain text, perhaps by being more concise, making some parts collapsible, or moving the background information and prediction market concepts into a separate article.

  ![image](https://github.com/user-attachments/assets/f388c209-168e-495f-a3d2-e6a7a53a5815)

  Quick and dirty gpt 4.1 1shot prompt to test different formatting:
  ![image](https://github.com/user-attachments/assets/15ed183b-875c-4db9-a204-2416f60952b1)

I think you can get an idea now of what I meant with my original comment, but I'm happy to share all the parts that I feel could be made more digestible for the users 🙌

rin-st commented on README.md:
```diff
@@ -595,7 +595,7 @@ There's one small step left, since we've added new state variables that need to
 Now you could head over to the oracle tab in the UI and report the outcome (after you watched the race of course:)) 🏎️ 🏁
 
-> 💡 Make sure you're connected with the correct oracle address — check `00_deploy_your_contract.ts` to find out which one is being used.
+> 💡 Make sure you're connected with the correct oracle address — check `00_deploy_your_contract.ts` to find out which one is being used. (Hint: Add the first Anvil account to your wallet to interact as the oracle or contract owner.)
```
Thanks for the update! One more nitpick: Since currently sre challenges works with Hardhat, it's better to change `Anvil` to `Hardhat`

phipsae commented:
@Pabl0cks thanks for the feedback! 
I’ve updated the front-end to match the style of the other challenges—let me know if I missed anything.
I also reworked the README to make it more engaging with less text.
Haven’t touched README.md.args.mjs yet—will update that once the main README is finalized.

Pabl0cks commented on README.md:
```diff
@@ -1,6 +1,1555 @@
+<details markdown='1'><summary>Why Prediction Markets Matter (click to expand)</summary>
+Prediction markets have been around for a long time, with records of **election betting on Wall Street dating back to 1884** (see [this Wikipedia page](https://en.wikipedia.org/wiki/Prediction_market)). On **Ethereum**, they've been a topic of interest for quite some time, but they never really took off, until now.
+
+Polymarket, a prediction market, is one of the most widely used blockchain applications today by both crypto natives and everyday users. It gained especially significant traction around the **U.S. elections.**
```
nit: I'd highlight **Polymarket** with bold

Pabl0cks commented on README.md:
```diff
@@ -1,6 +1,1555 @@
+
+The more people bet on a particular outcome, the **more expensive** that side becomes, while the opposite side gets **cheaper**. This **dynamic pricing mechanism** determines an implied **probability**, which can sometimes be more accurate than expert opinions, polls, or pundits.
+
+In this tutorial, we'll guide you through the fundamental Solidity functions and explore how a fully on-chain prediction market could be structured. Our focus will be on trading and liquidity provision. This decentralized prediction market enables anyone to place bets on a predefined question for a future event with the two outcomes "Yes" or "No".
```
This last paragraph doesn't feel like should be into the "Why Prediction Markets Matter" collapsable section, since is talking about our challenge.

Maybe we could merge the first part somehow in the first paragraph of the introduction (`In this tutorial, we'll guide you through the fundamental Solidity functions and explore how a fully on-chain prediction market could be structured`).

Perhaps we could remove the second part (`This decentralized prediction market enables anyone to place bets on a predefined question for a future event with the two outcomes "Yes" or "No"`) since we are going to explain it again in the next section

phipsae replied:
What do you think about just linking this article? 
https://medium.com/@philip.krause88/building-web3-prediction-markets-how-they-work-and-where-ai-takes-them-next-e87a6b3511e6

Pabl0cks replied:
> What do you think about just linking this article?
https://medium.com/@philip.krause88/building-web3-prediction-markets-how-they-work-and-where-ai-takes-them-next-e87a6b3511e6

SEO and usability wise, I think I would split this article into four smaller articles, each one targeting a different search intent and addressing a specific knowledge gap, and create them into SRE website instead of medium. 
I’d add a call to action with a link to the challenge at the end of each article, and I’d link the articles in the README wherever they are most relevant, so users can dig into them if they are interested in learning more. 

That would slim down our README and help with SEO (I think)

phipsae replied:
Okay, I kept it without the medium link. The chapters doesn't really fit exactly in there, since the article   is coherent. If required I can add later more text. 

I think the article like it is a great introduction for everyone who is interested in building a prediction market. So maybe in the SRE blog (you mentioned) if this is better for SEO?

Pabl0cks replied:
I think we need to publish articles that complement the main content (the challenge). Since it is published already in medium, I don't know how google is going to act in terms of duplicated content.

We can leave it like it is now, and maybe in the future you can prepare articles that complement the challenge content, or trim the main article in different articles as I suggested before. I'm preparing [articles for the initial challenges](https://github.com/BuidlGuidl/SpeedRunEthereum-v2/pull/169), maybe once I have a few of them, we can check which articles would be nice to have for this challenge.

Pabl0cks commented on README.md:
```diff
@@ -1,6 +1,1555 @@
+
+- But they also take on risk, since losing tokens become worthless
+
+That’s why popular platforms curate markets to ensure there’s enough interest and volume (see Polymarket docs).
```
Maybe we could add a link to the docs

Pabl0cks commented on README.md:
```diff
@@ -1,6 +1,1555 @@
+    <img src="extension/packages/nextjs/public/lp1.png" alt="ch-6-lp1" />
+</details>
+
+> ❗️In the current state your front-end is already implemented but the buttons of the different function will likely break since there is no implementation code within your smart contracts yet. But soon there will! 🙂
```
different function => different functions

Pabl0cks commented on README.md:
```diff
@@ -1,6 +1,1555 @@
+
+## **Checkpoint 2: 🔭 The Prediction Market Setup 🏠**
+
+Before we can deploy your prediction market smart contract, we need to set up the constructor and declare some critical variables — this is the foundation the whole market will run on.
```
your prediction => our prediction ?

Pabl0cks commented on README.md:
```diff
@@ -1,6 +1,1555 @@
+
+When you deploy the market, you’ll provide:
+
+- **🧙 `_oracle`** – the address that will later report the outcome
```
Do we miss the `_liquidityProvider` in this constructor setup?

Pabl0cks commented on README.md:
```diff
@@ -1,6 +1,1555 @@
+- **🏆 `s_ethCollateral`**: the total ETH backing the tokens — think of this as your prize pool
+- **💸 `s_lpTradingRevenue`** - tracks the fees earned from users buying/selling tokens — the LP’s reward
+
+> ❗️For easier testing, we set the oracle address to be the same as the liquidity provider during deployment (see `00_deploy_your_contract.ts`). Also, **double-check the parameter values** we're passing into the constructor, like `_question`, etc. (Hint: Add the first Hardhat account to your wallet to interact as the oracle or contract owner.)
```
Maybe we could hint to set the `liquidityProvider` /  `oracle` addresses in the `00_deploy_your_contract.ts` scripts instead, like this?

![image](https://github.com/user-attachments/assets/775347d2-04f7-4ab5-8071-ba277e867a24)

phipsae replied:
I think I would leave it like it is, since so they can run `yarn account:import` and have their account then as `liquidity provider` and `oracle` set. But let me change the readme a bit, maybe like:

Hint: Add the first Hardhat account to your wallet or add your own account to interact as the oracle or contract owner. You can manually set your address in the deployment script or run `yarn account:import`.

Pabl0cks commented on README.md:
````diff
@@ -1,6 +1,1555 @@
+yarn test --grep "Checkpoint2"
+```
+
+> 🚨 Before we deploy the contract we need to finish implementing the constructor in the next checkpoint 3.
````
supernit: checkpoint 3 => Checkpoint 3

Pabl0cks commented on README.md:
```diff
@@ -1,6 +1,1555 @@
+
+</details>
+
+Run the following command to check if you implement all variable and checks correctly.
```
Nit: "Run the following command to check if you have implemented all variables and checks correctly." would fit better here?

Pabl0cks commented on README.md:
```diff
@@ -1,6 +1,1555 @@
+
+$$initialYesProbability = 60$$
+
+But here’s the twist: there haven't been any trades yet. So how can the protocol incoproate he initial probability?
```
incoproate => incorporate

Pabl0cks commented on README.md:
```diff
@@ -1,6 +1,1555 @@
+   Without locking, a single "Yes" purchase pushes the probability to 100%:
+   $$\frac{1}{1 + 0} = 100\%$$
+2. ⚖️ Price stability – Token price depends on the current market probability:
+   $$
+   tokenPrice = initialTokenValue * marketProbability
+   $$
```
Is not getting the math format correctly

Pabl0cks commented on README.md:
```diff
@@ -1,6 +1,1555 @@
+
+</details>
+
+Run the following command to check if you implement all variable and checks correctly.
```
Should we write it like this here also? => "if you have implemented all variables"

Pabl0cks commented on README.md:
```diff
@@ -1,6 +1,1555 @@
+
+🧩 Uncomment everything below: `/// Checkpoint 3 ///`
+
+This function brings our values to the front end.
```
nit: front-end to be inline with the others?

Pabl0cks commented on README.md:
````diff
@@ -1,6 +1,1555 @@
+yarn test --grep "Checkpoint8"
+```
+
+And then run `yarn deploy` to test it in the front end and see how the probability changes.
````
nit: front-end to be inline with the others?

Pabl0cks commented on README.md:
```diff
@@ -1,6 +1,1555 @@
+
+Now that your market is live, it’s time to give the liquidity provider more control 💪
+
+In this checkpoint, you'll implement the addLiquidity and removeLiquidity functions — allowing the market owner to inject more ETH or pull it out, while automatically minting or burning outcome tokens to keep the balance.
```
Should we set `addLiquidity` and `removeLiquidity` like this?

Pabl0cks commented on README.md:
```diff
@@ -1,6 +1,1555 @@
+
+In this checkpoint, you'll implement the addLiquidity and removeLiquidity functions — allowing the market owner to inject more ETH or pull it out, while automatically minting or burning outcome tokens to keep the balance.
+
+To keep things simple, we’re only allowing the original market creator to call these functions. That’s already enforced with the onlyOwner modifier 🔒
```
onlyOwner => `onlyOwner` ?

Pabl0cks commented on README.md:
```diff
@@ -1,6 +1,1555 @@
+
+To do that, it will call a smart contract write function and pass in the final outcome — either "Yes" or "No".
+
+That outcome needs to be stored and remembered by the contracte.
```
contracte => contract

Pabl0cks commented on README.md:
```diff
@@ -1,6 +1,1555 @@
+
+- A `report()` function the oracle can call to finalize the outcome
+
+- A variable `s_winningToken` to store the winner (Yes or No token contract) and
```
We could remove the "and" at the end

Pabl0cks commented on README.md:
```diff
@@ -1,6 +1,1555 @@
+
+<details markdown='1'><summary>Question 3</summary>
+
+> How do make sure that the report function can only be called when the prediction is not reported (modifier)? Which other existing functions should also use the predictionNotReported modifier?
```
Missing the you/we in "How do make sure" or "How to make sure"

Pabl0cks commented on README.md:
```diff
@@ -1,6 +1,1555 @@
+
+Now head over to the oracle tab in the UI and report the outcome (after you watched the race of course:)) 🏎️ 🏁
+
+> 💡 Make sure you're connected with the correct oracle address — check `00_deploy_your_contract.ts` to find out which one is being used. (Hint: Add the first Hardhat account to your wallet to interact as the oracle or contract owner.)
```
If we finally change the hint to set your own address in the 00_deploy... in the previous hint, we should do the same here

Pabl0cks commented on README.md:
```diff
@@ -1,6 +1,1555 @@
+
+Welcome to the part of the prediction market, where price meets probability, and every trade shifts the odds. Let’s make it happen! 💥
+
+### 🛠️ Your Goal: Enbale trading
```
Enbale => Enable

Pabl0cks commented on README.md:
````diff
@@ -1,6 +1,1555 @@
+yarn test --grep "Checkpoint9"
+```
+
+Then run `yarn deploy` to test it on the frontend. Make sure to purchase some winning tokens beforehand and report the race. After that, you should be able to redeem your desired amount.
````
supernit: we should stick with frontend or front-end through all the challenge

Pabl0cks reviewed (commented):
> Looking forward to your feedback and thank you for taking your time!

Thanks @phipsae !! I like this format more, feels easier to consume! 

Added a few comments, mostly nitpicks or typos 🙏

phipsae commented:
@Pabl0cks thanks a lot for your valuable feedback! 

I updated the ReadMe accordingly. 

If we are ready to merge let me know, then I can squash the commits. 
I think we also need to change the branch name to `challenge-prediciton-markets`

Pabl0cks reviewed (approved):
Thx @phipsae !! Haven't re-read everything, but I think it looks good to merge!

Just if you could run a final AI grammar-spelling check, I think we missed some, example:

- "Futher" → "Further"
- "itself" → "themselves" (in "Prediction markets itself have been around...")
- "Therefor" → "Therefore"

> I think we also need to change the branch name to challenge-prediciton-markets

Yeah! I think Elliot updated over-collatelarized to include it too 🙏

phipsae commented:
Thx @Pabl0cks!! 

Typos fixed and commits squashed.

Pabl0cks commented:
> Typos fixed and commits squashed.

Thanks Phillip! We usually do the "Squash and merge", don't need to squash yourself next time 🙌

Let's merge this and fix whatever is left / we find during our first tests with users in a different PR 🙏
